### PR TITLE
Unadvertise now

### PR DESCRIPTION
**Pull request by anson627**
Close peer and delete it from peer heap when unadvertise
Not destroy hyperbahn client so that it can re-advertise
@ShanniLi @Raynos @kriskowal 
cc: @davewhat @vipulaneja

### CODE REVIEW

File: node/hyperbahn/index.js [hunk 1]
```diff
@@ -362,12 +362,22 @@ function advertise(opts) {
...
     var self = this;
-    self.destroy();
+    self.sendRequest(opts, 'unad', unadvertiseInternalCb);
+    timers.clearTimeout(self._advertisementTimer);
+    timers.clearTimeout(self.advertisementTimeoutTimer);
     self.latestAdvertisementResult = null;
     self.state = States.UNADVERTISED;
     self.emit('unadvertised');
+    function unadvertiseInternalCb(error, result) {
+        self.logger.debug('HyperbahnClient: unadvertisement complete', {
```
Comment (ShanniLi) [inline node/hyperbahn/index.js]: please use logger.warn

File: node/hyperbahn/index.js [hunk 2]
```diff
@@ -387,7 +397,7 @@ HyperbahnClient.prototype.getHealthyRetryTime = function getHealthyRetryTime() {
 };
 
 HyperbahnClient.prototype.advertiseAgain =
-function advertiseAgain(delay) {
+    function advertiseAgain(delay) {
```
Comment (ShanniLi) [inline node/hyperbahn/index.js]: why space here?

File: node/test/hyperbahn/unadvertise.js [hunk 3]
```diff
@@ -35,7 +35,6 @@ if (require.main === module) {
 function runTests(HyperbahnCluster) {
     HyperbahnCluster.test('advertise and forward', {
         size: 5,
-        servicePurgePeriod: 50
```
Comment (ShanniLi) [inline node/test/hyperbahn/unadvertise.js]: Please keep this test and create a new one in this file for unadvertise endpoint.

File: node/test/hyperbahn/unadvertise.js [hunk 4]
```diff
@@ -75,8 +74,7 @@ function runTests(HyperbahnCluster) {
     });
 
     HyperbahnCluster.test('advertise, unadvertise and forward', {
-        size: 5,
-        servicePurgePeriod: 50
```
Comment (ShanniLi) [inline node/test/hyperbahn/unadvertise.js]: Please keep this test and create a new one in this file for unadvertise endpoint.

File: node/hyperbahn/handler.js [hunk 5]
```diff
@@ -178,9 +195,20 @@ function sendAdvertise(services, options, callback) {
 
     res: {}
 */
+
 HyperbahnHandler.prototype.handleRelayAdvertise =
 function handleRelayAdvertise(self, req, arg2, arg3, cb) {
-    /*eslint max-params: [2, 5]*/
+    self.handleRelay(req, arg2, arg3, cb, self.advertise);
```
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: This is generally bad. The reason it works here is self.advertise doesn't use this pointer.

File: node/hyperbahn/index.js (hunk 1 above)
Comment (anson627) [inline node/hyperbahn/index.js]: you mean warn on error, otherwise debug?

File: node/hyperbahn/index.js (hunk 2 above)
Comment (anson627) [inline node/hyperbahn/index.js]: will remove

File: node/test/hyperbahn/unadvertise.js (hunk 3 above)
Comment (anson627) [inline node/test/hyperbahn/unadvertise.js]: will do

File: node/test/hyperbahn/unadvertise.js (hunk 4 above)
Comment (anson627) [inline node/test/hyperbahn/unadvertise.js]: will do

File: node/hyperbahn/handler.js [hunk 6]
```diff
@@ -296,3 +324,8 @@ HyperbahnHandler.prototype.advertise =
 function advertise(service) {
     throw new Error('not implemented');
 };
+
+HyperbahnHandler.prototype.unadvertise =
```
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: This should work:
self.channel.topChannel.handler.removeServicePeer(service.serviceName, service.hostPort);

Comment (anson627) [inline node/hyperbahn/handler.js]: oh, so you can get top channel from sub channel? let me try.

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (anson627) [inline node/hyperbahn/handler.js]: I don't understand the purpose for passing self as a parameter.

File: node/hyperbahn/index.js (hunk 1 above)
Comment (ShanniLi) [inline node/hyperbahn/index.js]: Oh, I missed this this for all cases. You may want to do logger.warn for error cases.

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: in this case, when func is called, "this" pointer will be null. When you do callback binding, the first parameter is "this".

File: node/hyperbahn/index.js (hunk 1 above)
Comment (anson627) [inline node/hyperbahn/index.js]: will do

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (anson627) [inline node/hyperbahn/handler.js]: I see, what do you suggest doing here? I want to share most of the code for advertise and unadvertise.

File: node/hyperbahn/handler.js (hunk 6 above)
Comment (anson627) [inline node/hyperbahn/handler.js]: it works.

File: node/hyperbahn/handler.js (hunk 5 above)
Comment (ShanniLi) [inline node/hyperbahn/handler.js]: self.advertise.bind(self) is a correct way, but @Raynos said it is not efficient.

Another way is to pass in the endpoint name and do an if-else in self.handleRelay

**Comment by ShanniLi** [comment]
:+1: 
you will also need to update hyperbahn for .advertise

**Comment by anson627** [comment]
@ShanniLi I can just remove the advertise function override from hyperbahn, right?

**Comment by ShanniLi** [comment]
yes

File: node/hyperbahn/handler.js [hunk 7]
```diff
@@ -294,5 +329,12 @@ function logError(err, opts, response) {
 
 HyperbahnHandler.prototype.advertise =
 function advertise(service) {
-    throw new Error('not implemented');
+    var self = this;
+    self.channel.topChannel.handler.refreshServicePeer(service.serviceName, service.hostPort);
```
Comment (Raynos) [inline node/hyperbahn/handler.js]: This is not ideal; Would be better to do `self.serviceProxy.refreshServicePeer()`

File: node/hyperbahn/index.js [hunk 8]
```diff
@@ -362,12 +362,29 @@ function advertise(opts) {
...
+    function unadvertiseInternalCb(error, result) {
+        if (error) {
+            self.logger.warn('HyperbahnClient: unadvertisement failure', {
+                error: error,
+                serviceName: self.serviceName,
+                hostPort: self.hostPort
+            });
+            return;
+        }
+        self.logger.debug('HyperbahnClient: unadvertisement success', {
```
Comment (Raynos) [inline node/hyperbahn/index.js]: No `debug()`.

**Comment by Raynos** [comment]
@anson627 Let's open a new PR to remove the `logger.debug()`.

We generally do not want any debug logs in tchannel. Logging that you did what you were told is generally not very useful,

**Comment by Raynos** [comment]
Other wise looks good.

File: node/hyperbahn/index.js (hunk 8 above)
Comment (anson627) [inline node/hyperbahn/index.js]: Will do. Good to know our logging strategy.

File: node/hyperbahn/handler.js (hunk 7 above)
Comment (anson627) [inline node/hyperbahn/handler.js]: will do in my next change.

Comment (anson627) [inline node/hyperbahn/handler.js]: HyperbahnHandler doesn't have reference to ServiceDispatchHandler(service_proxy)

**Pull request merged**